presto-plan-checker-router-plugin/src/main/java/com/facebook/presto/router/scheduler/PlanCheckerRouterPluginPrestoClient.java

```diff
@@ -0,0 +1,243 @@
+    private static final JsonCodec<SqlFunctionId> SQL_FUNCTION_ID_JSON_CODEC = jsonCodec(SqlFunctionId.class);
+    private static final JsonCodec<SqlInvokedFunction> SQL_INVOKED_FUNCTION_JSON_CODEC = jsonCodec(SqlInvokedFunction.class);
+    private static final String ANALYZE_CALL = "EXPLAIN (TYPE DISTRIBUTED) ";
+    private static final CounterStat javaClusterRedirectRequests = new CounterStat();
+    private static final CounterStat nativeClusterRedirectRequests = new CounterStat();
+
+    private final OkHttpClient httpClient = new OkHttpClient();
+    private final URI planCheckerClusterURI;
+    private final URI javaRouterURI;
+    private final URI nativeRouterURI;
+
+    public PlanCheckerRouterPluginPrestoClient(URI planCheckerClusterURI, URI javaRouterURI, URI nativeRouterURI)
+    {
+        this.planCheckerClusterURI = planCheckerClusterURI;
+        this.javaRouterURI = javaRouterURI;
+        this.nativeRouterURI = nativeRouterURI;
+    }
+
+    public Optional<URI> getCompatibleClusterURI(HttpServletRequest httpServletRequest, String statement)
+    {
+        String newSql = ANALYZE_CALL + statement;
+        ClientSession clientSession = parseHeadersToClientSession(httpServletRequest);
+        boolean isNativeCompatible = true;
+        // submit initial query
+        try (StatementClient client = newStatementClient(httpClient, clientSession, newSql)) {
+            // read query output
+            while (client.isRunning()) {
+                log.debug((client.currentData().toString()));
+
+                if (!client.advance()) {
+                    break;
+                }
+            }
+
+            // verify final state
+            if (client.isClientAborted()) {
+                throw new IllegalStateException("Query aborted by user");
+            }
+
+            if (client.isClientError()) {
+                throw new IllegalStateException("Query is gone (server restarted?)");
+            }
+
+            verify(client.isFinished());
+            QueryError resultsError = client.finalStatusInfo().getError();
+            if (resultsError != null) {
+                if (ErrorType.valueOf(resultsError.getErrorType()) == USER_ERROR) {
+                    throw new PrestoException(
+                            () -> new ErrorCode(
+                                    resultsError.getErrorCode(),
+                                    resultsError.getErrorName(),
+                                    USER_ERROR,
+                                    resultsError.isRetriable()),
+                            newQueryResults(client));
+                }
+                isNativeCompatible = false;
+                log.info(resultsError.getMessage());
+            }
+        }
+
+        if (isNativeCompatible) {
+            log.debug("Native compatible, routing to native-clusters router: [%s]", nativeRouterURI);
+            nativeClusterRedirectRequests.update(1);
```

Comment:
nit : `nativeClusterRedirectRequests.update(1L)` you can use a native long constant